### PR TITLE
feat: add support for templating

### PR DESCRIPTION
**Pull request by tuanddd**

### CODE REVIEW

File: packages/search-ui/src/Results/components/Result/index.tsx [hunk 1]
```diff
@@ -7,18 +7,26 @@ import {
...
 import { useTranslation } from 'react-i18next';
 
 import { useSearchUIContext } from '../../../ContextProvider';
 import { useClickTracking } from '../../../hooks';
+import { Template } from '../../types';
 import useResultStyles from './styles';
 import { ResultProps } from './types';
 
+export function checkValidTemplate(template?: Template): template is Template {
+  return !isNullOrUndefined(template) && !isEmptyObject(template) && !isEmpty(template?.html) && !isEmpty(template.css);
```
Comment (zlatanpham) [inline packages/search-ui/src/Results/components/Result/index.tsx:27]: Probably, it's ok to not pass the CSS?

File: docs/pages/search-ui/results.mdx [hunk 2]
```diff
@@ -86,6 +86,83 @@ function Example() {
...
+    html: `
+<a href="{{url}}">
+  <article class="relative p-3 rounded-md border border-gray-300 inline-block max-w-xs">
+     <img class="mx-auto" src="{{image}}" />
+     <p class="font-medium text-gray-700 mt-4">{{title}}</p>
+     <p class="text-xs text-gray-400 mt-1">{{subtitle}}</p>
+     <p class="text-gray-500 mt-2 text-sm">{{description}}</p>
+     <div class="flex mt-8 justify-between">
+       <span class="font-medium text-gray-700">\${{price}}</span>
+       {{#if true}}
```
Comment (zlatanpham) [inline docs/pages/search-ui/results.mdx:113]: It doesn't look like a good example to me. Can we use a field check here?

File: packages/search-ui/src/Results/components/Result/index.tsx [hunk 3]
```diff
@@ -209,6 +218,17 @@ const Result = React.memo(
...
+    if (checkValidTemplate(template)) {
+      try {
+        const handlebarTemplate = Handlebars.compile(template.html);
+        const rendered = handlebarTemplate(values);
+        // eslint-disable-next-line react/no-danger
+        return <div dangerouslySetInnerHTML={{ __html: rendered }} />;
+      } catch (error) {
+        return null;
+      }
+    }
```
Comment (zlatanpham) [inline packages/search-ui/src/Results/components/Result/index.tsx:230]: I can see we have a lot of calculations upper the block. Probably, we want to early return it for a sake of performance?

File: packages/search-ui/src/Results/index.tsx [hunk 4]
```diff
@@ -81,12 +83,32 @@ const Results = (props: ResultsProps) => {
...
+      return <Message title={t('common:error')} body={body} showReset />;
+    }
+  }
+
   return (
     <ResizeObserver
       onResize={(rect) => setWidth(rect.width)}
       css={[styles.container, stylesProp]}
       className={customClassNames.results?.container}
     >
```
Comment (zlatanpham) [inline packages/search-ui/src/Results/index.tsx:103]: I'm wondering`ResizeObserver` is needed if the results are rendered via a template. I think we probably need to have a container wrapper here. We can have some default CSS for flex display and spacing but we also need to provide an ability for users to input CSS classNames. What do you think?

File: docs/pages/search-ui/results.mdx [hunk 5]
```diff
@@ -86,13 +86,92 @@ function Example() {
 }
 ```
 
+### Custom templating
```
Comment (zlatanpham) [inline docs/pages/search-ui/results.mdx:89]: The example looks off to me. Can we have a grid display for it?

image

File: packages/search-ui/src/Results/components/TemplateResults/index.tsx [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+            as={resultContainerTemplateElement}
+          />
+        ))}
+      </div>
+    );
+  } catch (e) {
+    console.error(e);
+    const body = t('errors:template');
+    return <Message title={t('common:error')} body={body} showReset />;
+  }
```
Comment (zlatanpham) [inline packages/search-ui/src/Results/components/TemplateResults/index.tsx:52]: It's uncommon to handle rendering inside `try-catch`, I guess. Do you think if using [ErrorBoundary](https://reactjs.org/docs/error-boundaries.html) could make more sense?

**Review by zlatanpham — COMMENTED** [review]
I've left 2 more comments. @wwalser feel free to leave your words too.

File: docs/pages/search-ui/results.mdx (hunk 5 above)
Comment (tuanddd) [reply docs/pages/search-ui/results.mdx:89]: Will do

File: packages/search-ui/src/Results/components/TemplateResults/index.tsx (hunk 6 above)
Comment (tuanddd) [reply packages/search-ui/src/Results/components/TemplateResults/index.tsx:52]: Got it

File: packages/search-ui/src/Results/components/TemplateResults/index.tsx [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+  const { results, template, resultContainerTemplateElement } = props;
+  try {
+    // Get the keys of result
+    const keys = Array.from(
+      results?.reduce((acc, cur) => {
+        Object.keys(cur).forEach((k) => acc.add(k));
+        return acc;
+      }, new Set<string>()),
+    );
+    const render = compile(template.html, { async: false, props: keys });
```
Comment (wwalser) [inline packages/search-ui/src/Results/components/TemplateResults/index.tsx:24]: Should memoize the call to compile when `template.html` is the same on successive calls.

File: docs/pages/search-ui/results.mdx [hunk 8]
```diff
@@ -86,13 +86,92 @@ function Example() {
...
+| `showVariantImage`               | `boolean`                                    | `false`   | Sets to show variant images                                                                                                     |
+| `imageAspectRatio`               | `number` \| `{ list: number, grid: number}`  | `1`       | The aspect ratio to force images to render to.                                                                                  |
+| `imageObjectFit`                 | `string` \| `{ list: string, grid: string }` | `contain` | The aspect ratio to force images to render to. Valid `objectFit` values can be found in [image props](/components/image#props). |
+| `template`                       | `{html: string, css: string}`                |           | The template code use to render a single result item.                                                                           |
+| `resultContainerTemplateElement` | `"div" \| "p" \| ...`                        | `"div"`   | The element used to render the result container.                                                                                |
```
Comment (wwalser) [inline docs/pages/search-ui/results.mdx:177]: This is purely to get around the fact that `dangerouslySetInnerHTML` requires some wrapper, yes?

I think the text should read:
> The html element which will wrap each `resultTemplate`.

File: docs/pages/search-ui/results.mdx [hunk 9]
```diff
@@ -86,13 +86,92 @@ function Example() {
...
+| `showImage`                      | `boolean`                                    | `true`    | Sets to show images                                                                                                             |
+| `showVariantImage`               | `boolean`                                    | `false`   | Sets to show variant images                                                                                                     |
+| `imageAspectRatio`               | `number` \| `{ list: number, grid: number}`  | `1`       | The aspect ratio to force images to render to.                                                                                  |
+| `imageObjectFit`                 | `string` \| `{ list: string, grid: string }` | `contain` | The aspect ratio to force images to render to. Valid `objectFit` values can be found in [image props](/components/image#props). |
+| `template`                       | `{html: string, css: string}`                |           | The template code use to render a single result item.                                                                           |
```
Comment (wwalser) [inline docs/pages/search-ui/results.mdx:176]: Can we rename this to `resultTemplate` in case we discover a need to template more things in the future?

**Review by wwalser — APPROVED** [review]
Approving assuming you can handle requested changes before landing 👍
Thanks! This looks great.

**Comment by tuanddd** [comment]
Thanks @wwalser, I'll make sure to resolve all the comments.

**Review by zlatanpham — APPROVED** [review]

**Pull request merged**